app/services/bill-runs/cancel/delete-bill-run.service.js

```diff
@@ -0,0 +1,229 @@
+const { db } = require('../../../../db/db.js')
+const ChargingModuleDeleteBillRunRequest = require('../../../requests/charging-module/delete-bill-run.request.js')
+const { calculateAndLogTimeTaken } = require('../../../lib/general.lib.js')
+const ReviewChargeVersionModel = require('../../../models/review-charge-version.model.js')
+const ReviewLicenceModel = require('../../../models/review-licence.model.js')
+const ReviewReturnModel = require('../../../models/review-return.model.js')
+
+/**
+ * Deletes a bill run, all its associated records and the its match in the Charging Module API
+ *
+ * We first send a request to the Charging Module API to delete the bill run there. Then we delete all details of the
+ * bill run from our tables.
+ *
+ * Where possible we do these in parallel to speed up the process. We also are not to fussed about errors. For example,
+ * if the CHA returns an error we don't stop the process as an orphaned CHA bill run won't impact us.
+ *
+ * The final thing worth noting that due to the number of constraints placed on the legacy tables deleting records can
+ * take some time. For example, an Anglian annual bill run in production will take more than 30 mins to delete
+ * everything.
+ *
+ * @param {module:BillRunModule} billRun - The bill run to be deleted
+ */
+async function go (billRun) {
+  try {
+    const startTime = process.hrtime.bigint()
+
+    const { id: billRunId, externalId } = billRun
+
+    // Within the scope of this method we await all 3 processes to complete and then log the time taken. For a small
+    // supplementary bill run the request to the charge module might be the one that takes the longest. In the case of
+    // an annual bill run it will be removing the bill run. Either way, all 3 processes can be run in parallel as none
+    // of them have a dependence on the
+    // other
+    const results = await Promise.allSettled([
+      // If the Charging Module errors whilst doing this it shouldn't block us carrying on with deleting the bill run on
+      // our side. It just means the the CHA will be storing an 'orphaned' bill run that will never get sent.
+      ChargingModuleDeleteBillRunRequest.send(externalId),
+      // We can be deleting these records whilst getting on with deleting the other things. But should it fail we'll
+      // just be left with orphaned review results. As long as it's an incidental occurrence this wouldn't be a problem.
+      _deleteReviewResults(billRunId),
+      // Finally we get on with deleting the billing records which in some cases (Anglian annual bill run) can take more
+      // than 30 mins!
+      _deleteBillingRecords(billRunId)
+    ])
+
+    _logResult(startTime, billRun, results)
+  } catch (error) {
+    global.GlobalNotifier.omfg('Delete bill run failed', billRun, error)
+  }
+}
+
+/**
+ * These deals with all the tables that may be populated for a bill run. Because we have to deal with bills created
```

Comment:
this should start with 'this' I think?

Reply:
💯 

But I'm about to do exactly the same refactor for sending a bill run. So, I promise 😇 I'll fix it in that PR rather than wasting you lovely approval!